src/Symfony/Bundle/FrameworkBundle/Controller/RedirectController.php

```diff
@@ -54,15 +54,16 @@ public function __construct(UrlGeneratorInterface $router = null, int $httpPort
      *
      * @throws HttpException In case the route name is empty
      */
-    public function redirectAction(Request $request, string $route, bool $permanent = false, $ignoreAttributes = false, bool $keepRequestMethod = false): Response
+    public function redirectAction(Request $request, string $route, bool $permanent = false, $ignoreAttributes = false, bool $keepRequestMethod = false, bool $keepQueryParams = false): Response
     {
         if ('' == $route) {
             throw new HttpException($permanent ? 410 : 404);
         }
 
         $attributes = array();
         if (false === $ignoreAttributes || is_array($ignoreAttributes)) {
-            $attributes = $request->attributes->get('_route_params');
+            $routeParams = $request->attributes->get('_route_params');
+            $attributes = $keepQueryParams ? \array_merge($request->query->all(), $routeParams) : $routeParams;
```

Comment:
I suggest adding a test ensuring that attributes coming from `_route_params` still win over attributes coming from the query string when merging them together.